### PR TITLE
Add tests for item_pocket and pocket_data

### PR DESCRIPTION
**Pull request by wapcaplet**
#### Summary

SUMMARY: Infrastructure "Add tests for item_pocket and pocket_data"

#### Purpose of change

To begin fleshing out tests dedicated to item pockets and pocket data, fundamental parts of the nested container system.

#### Describe the solution

Add test data to `data/mods/TEST_DATA` and test cases to `tests/pocket_test.cpp`. Add new flags "max_item_length" and "max_item_volume" to `doc/JSON_INFO.md` and reformat the pocket data section for slightly better readability. Adds a convenience function `item_pocket::airtight()` to mirror `item_pocket::watertight()`.

Tests included:

- pocket with max_contains_volume
  - with ammo_restriction, ignored in favor of the ammo type volume and count
  - without ammo_restriction, same as volume_capacity
- pocket with max_item_volume
  - can contain items that fit through the opening
  - can contain soft items, even if too large for the opening
  - can contain liquids
- pocket with max_item_length
  - can contain items with longest_side that length or shorter
  - cannot contain items with longest_side longer than that
- magazine with ammo restriction
  - can contain the appropriate ammo, unless it is full already
  - cannot contain wrong ammo types
  - cannot contain things that aren't ammo
- pocket with a flag restriction (like BELT_CLIP)
  - can contain items with matching flag
  - unless they are less than min_item_volume
  - or they are greater than max_contains_volume
- watertight pocket
  - can contain non-liquids
  - can contain liquids
  - unless a different liquid is already in it
  - or a non-liquid is already in it
- non-watertight pocket
  - cannot contain liquids
  - can contain non-liquids
- containing gases
  - airtight pocket can contain gas
  - non-airtight pocket cannot contain gas
- non-rigid pocket
  - item_size_modifier grows in relation to contained items
  - can contain items when empty or partly filled
  - cannot contain more items once full
- rigid pocket
  - item_size_modifier is always 0
  - can contain items when empty
  - cannot contain more items once full
- holster pocket
  - can contain a item
  - unless the item is too heavy
  - or the item is too big
  - or the holster already has an item in it
- corpses can contain anything

#### Describe alternatives you've considered

I considered getting a jump on these a few weeks ago, but wanted to wait for some of the dust around nested containers to settle first. There are still some particulates floating in the air, but it feels like the right time to begin now.

There are some TODO items left in `pocket_test.cpp` as potentially low-hanging fruit for future contributors.

#### Testing

```
$ tests/cata_test [pocket]
All tests passed (320 assertions in 10 test cases)
The test took 0.002 seconds
```

#### Additional context

Finished and rebased into a few nice commits now, and all tests pass.

### CODE REVIEW

File: tests/pocket_test.cpp [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+                                   item_pocket::contain_code::ERR_NO_SPACE );
+        }
+    }
+}
+
+TEST_CASE( "magazine with ammo restriction", "[pocket][magazine][ammo_restriction]" )
+{
+    // Generic magazine with plenty of capacity for almost any ammo
+    pocket_data data_mag( item_pocket::pocket_type::MAGAZINE );
+    data_mag.volume_capacity = 10_liter;
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:107]: i recommend you actually test with these set at default values/ 0, since that's how they mostly get loaded into the json (itemfactory is just being special for now)

Comment (wapcaplet) [reply tests/pocket_test.cpp]: Thanks, and confirmed - even explicit 0 capacity works, so I've included that in the last revision with an explanaion.

**Comment by wapcaplet** [comment]
Travis CI build is failing, but it looks unrelated to any code changes in this PR:

```
src/npc.cpp:692 [virtual void npc::setpos(const tripoint&)] could not find npc  on its old overmap
Treating result as failure due to error logged during tests.
```

File: data/mods/TEST_DATA/items.json [hunk 2]
```diff
@@ -516,11 +557,12 @@
     "description": "A poorly balanced sword for test purposes",
     "category": "spare_parts",
     "weight": "1250 g",
+    "volume": "1 L",
+    "longest_side": "100 cm",
```
Comment (KorGgenT) [inline data/mods/TEST_DATA/items.json:561]:
```suggestion
    "longest_side": "1 meter",
```
i remember what you said about tests doubling as documentation. i also like having the largest denomination of unit possible.

File: data/mods/TEST_DATA/items.json [hunk 3]
```diff
@@ -533,11 +575,12 @@
     "description": "A sword for test purposes",
     "category": "spare_parts",
     "weight": "1250 g",
+    "volume": "1 L",
+    "longest_side": "100 cm",
```
Comment (KorGgenT) [inline data/mods/TEST_DATA/items.json:579]:
```suggestion
    "longest_side": "1 meter",
```

File: data/mods/TEST_DATA/items.json [hunk 4]
```diff
@@ -550,14 +593,118 @@
     "description": "A well-balanced sword for test purposes",
     "category": "spare_parts",
     "weight": "1250 g",
+    "volume": "1 L",
+    "longest_side": "100 cm",
```
Comment (KorGgenT) [inline data/mods/TEST_DATA/items.json:597]:
```suggestion
    "longest_side": "1 meter",
```

File: data/mods/TEST_DATA/items.json [hunk 5]
```diff
@@ -550,14 +593,118 @@
...
+  {
+    "id": "test_gas",
+    "type": "AMMO",
+    "ammo_type": "test_gas",
+    "category": "chems",
+    "name": { "str_sp": "test gas" },
+    "description": "Some mysterious substance in the form of a gas.  Only for testing, do not inhale!",
+    "phase": "gas",
+    "weight": 250,
+    "volume": 1,
```
Comment (KorGgenT) [inline data/mods/TEST_DATA/items.json:705]:
```suggestion
    "volume": "250 ml",
```

File: data/mods/TEST_DATA/items.json [hunk 6]
```diff
@@ -550,14 +593,118 @@
...
+  },
+  {
+    "id": "test_gas",
+    "type": "AMMO",
+    "ammo_type": "test_gas",
+    "category": "chems",
+    "name": { "str_sp": "test gas" },
+    "description": "Some mysterious substance in the form of a gas.  Only for testing, do not inhale!",
+    "phase": "gas",
+    "weight": 250,
```
Comment (KorGgenT) [inline data/mods/TEST_DATA/items.json:704]:
```suggestion
    "weight": "250 g",
```

File: data/mods/TEST_DATA/items.json [hunk 7]
```diff
@@ -550,14 +593,118 @@
...
+  {
+    "id": "test_liquid",
+    "type": "COMESTIBLE",
+    "comestible_type": "DRINK",
+    "category": "food",
+    "name": { "str_sp": "test liquid" },
+    "description": "No clue what it's made of, but it's definitely liquid.  Only for testing, do not drink!",
+    "phase": "liquid",
+    "weight": 250,
+    "volume": 1,
```
Comment (KorGgenT) [inline data/mods/TEST_DATA/items.json:685]:
```suggestion
    "volume": "250 ml",
```

File: data/mods/TEST_DATA/items.json [hunk 8]
```diff
@@ -550,14 +593,118 @@
...
+  },
+  {
+    "id": "test_liquid",
+    "type": "COMESTIBLE",
+    "comestible_type": "DRINK",
+    "category": "food",
+    "name": { "str_sp": "test liquid" },
+    "description": "No clue what it's made of, but it's definitely liquid.  Only for testing, do not drink!",
+    "phase": "liquid",
+    "weight": 250,
```
Comment (KorGgenT) [inline data/mods/TEST_DATA/items.json:684]:
```suggestion
    "weight": "250 g",
```

File: doc/JSON_INFO.md [hunk 9]
```diff
@@ -1883,23 +1883,26 @@ Any Item can be a container. To add the ability to contain things to an item, yo
...
-    "sealed_data": { "spoil_multiplier": 0.0 } // have anything in sealed_data means the pocket cannot be resealed. Additionally, the sealed version of the pocket will override the unsealed version of the same datatype.
+    "pocket_type": "CONTAINER",       // Typical container pocket. Pockets can also be MAGAZINE.
+    "max_contains_volume": mandatory, // Maximum volume this pocket can hold, totaled among all contained items.  For example "2 L" or "2000 ml" would hold two liters of items.
+    "max_contains_weight": mandatory, // Maximum weight this pocket can hold, totaled among all container items.  For example "6 kg" is about enough to contain a bowling ball.
+    "min_item_volume": "0 ml",        // Minimum volume of item that can be placed into this pocket.  Items smaller than this cannot be placed in the pocket.
+    "max_item_volume": "0 ml",        // Maximum volume of item that can fit through the opening into this pocket.  For example, a 2-liter bottle has a "17 ml" opening.
+    "max_item_length": "0 mm",        // Maximum length of items that can fit in this pocket, by their longest_side.
```
Comment (KorGgenT) [inline doc/JSON_INFO.md:1891]: i think it's worth mentioning in the documentation that the mandatory member doesn't default to zero, but the default is calculated based on its volume. in this case, it's cbrt( volume ) * sqrt( 2 )

File: tests/pocket_test.cpp [hunk 10]
```diff
@@ -0,0 +1,832 @@
...
+    CHECK_FALSE( rate_can.success() );
+    CHECK( rate_can.str() == expect_reason );
+    CHECK( rate_can.value() == expect_code );
+}
+
+// Call pocket.insert_item( it ) and expect it to be successful (contain_code::SUCCESS)
+static void expect_can_insert( item_pocket &pocket, item &it )
+{
+    CAPTURE( it.tname() );
+    ret_val<item_pocket::contain_code> rate_can = pocket.can_contain( it );
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:62]: you aren't calling insert_item like the comment suggests

File: tests/pocket_test.cpp [hunk 11]
```diff
@@ -0,0 +1,832 @@
...
+// -  open_container
+// -  sealed_data
+// -  moves
+
+// FIXME: The failure messaging from these helper functions is pretty terrible, due to the multiple
+// CHECKs in each function. Capturing it.tname() helps only slightly. The `rate_can.value()` is an
+// integer, so it is difficult to see when a failure reason is returned instead of success.
+
+// Expect pocket.can_contain( it ) to be successful (contain_code::SUCCESS)
+static void expect_can_contain( item_pocket &pocket, item &it )
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:38]:
```suggestion
static void expect_can_contain( const item_pocket &pocket, item &it )
```
can_contain() is a const function, might as well make this const too

File: tests/pocket_test.cpp [hunk 12]
```diff
@@ -0,0 +1,832 @@
...
+    CAPTURE( it.tname() );
+    ret_val<item_pocket::contain_code> rate_can = pocket.can_contain( it );
+    CHECK( rate_can.success() );
+    CHECK( rate_can.str().empty() );
+    CHECK( rate_can.value() == item_pocket::contain_code::SUCCESS );
+}
+
+// Expect pocket.can_contain( it ) to fail, with an expected reason and contain_code
+static void expect_cannot_contain( item_pocket &pocket, item &it, std::string expect_reason,
+                                   item_pocket::contain_code expect_code )
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:49]:
```suggestion
static void expect_cannot_contain( const item_pocket &pocket, item &it, const std::string &expect_reason,
                                   item_pocket::contain_code expect_code )
```

File: tests/pocket_test.cpp [hunk 13]
```diff
@@ -0,0 +1,832 @@
...
+
+// FIXME: The failure messaging from these helper functions is pretty terrible, due to the multiple
+// CHECKs in each function. Capturing it.tname() helps only slightly. The `rate_can.value()` is an
+// integer, so it is difficult to see when a failure reason is returned instead of success.
+
+// Expect pocket.can_contain( it ) to be successful (contain_code::SUCCESS)
+static void expect_can_contain( item_pocket &pocket, item &it )
+{
+    CAPTURE( it.tname() );
+    ret_val<item_pocket::contain_code> rate_can = pocket.can_contain( it );
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:41]:
```suggestion
    const ret_val<item_pocket::contain_code> rate_can = pocket.can_contain( it );
```
const where possible. pretty sure this is negligible, but i don't see any harm in doing this - the same for the others

File: tests/pocket_test.cpp [hunk 14]
```diff
@@ -0,0 +1,832 @@
...
+    CHECK( rate_can.success() );
+    CHECK( rate_can.str().empty() );
+    CHECK( rate_can.value() == item_pocket::contain_code::SUCCESS );
+}
+
+// Call pocket.insert_item( it ) and expect it to fail, with an expected reason and contain_code
+static void expect_cannot_insert( item_pocket &pocket, item &it, std::string expect_reason,
+                                  item_pocket::contain_code expect_code )
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:70]:
```suggestion
static void expect_cannot_insert( item_pocket &pocket, item &it, const std::string &expect_reason,
                                  item_pocket::contain_code expect_code )
```
passing a const reference to a string is much faster, though it doesn't matter for int and other primitive types, which an enum is

File: tests/pocket_test.cpp [hunk 15]
```diff
@@ -0,0 +1,832 @@
...
+{
+    CAPTURE( it.tname() );
+    ret_val<item_pocket::contain_code> rate_can = pocket.can_contain( it );
+    CHECK_FALSE( rate_can.success() );
+    CHECK( rate_can.str() == expect_reason );
+    CHECK( rate_can.value() == expect_code );
+}
+
+// Call pocket.insert_item( it ) and expect it to be successful (contain_code::SUCCESS)
+static void expect_can_insert( item_pocket &pocket, item &it )
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:59]:
```suggestion
static void expect_can_insert( item_pocket &pocket, const item &it )
```

File: tests/pocket_test.cpp [hunk 16]
```diff
@@ -0,0 +1,832 @@
...
+    GIVEN( "large watertight pocket with a narrow opening" ) {
+        item_pocket pocket_jug( &data_jug );
+
+        THEN( "it can contain liquids" ) {
+            expect_can_contain( pocket_jug, water );
+        }
+        THEN( "it can contain solid items that fit through the opening" ) {
+            expect_can_contain( pocket_jug, screwdriver );
+        }
+        THEN( "it can contain soft items larger than the opening" ) {
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:174]: following your other conventions here i expected an assertion that a rag is larger in volume than the mouth opening

File: tests/pocket_test.cpp [hunk 17]
```diff
@@ -0,0 +1,832 @@
...
+        // 9mm ammo is 50 rounds per 250ml (or 200 rounds per liter), so this ammo box
+        // should be exactly 1 liter in size, so it can contain this much ammo.
+        data_ammo_box.ammo_restriction.emplace( ammotype( "9mm" ), 200 );
+        REQUIRE_FALSE( data_ammo_box.ammo_restriction.empty() );
+
+        // And because actual volume is derived from ammo needs, this volume should be ignored.
+        data_ammo_box.volume_capacity = 1_ml;
+
+        THEN( "max_contains_volume is based on the pocket's ammo type" ) {
+            CHECK( data_ammo_box.max_contains_volume() == 1000_ml );
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:226]:
```suggestion
            CHECK( data_ammo_box.max_contains_volume() == 1_liter );
```
any reason this isn't just 1 liter?

File: tests/pocket_test.cpp [hunk 18]
```diff
@@ -0,0 +1,832 @@
...
+        //
+        const int full_clip_qty = 10;
+        data_mag.ammo_restriction.emplace( ammotype( "9mm" ), full_clip_qty );
+        item_pocket pocket_mag( &data_mag );
+
+        WHEN( "it does not already contain any ammo" ) {
+            REQUIRE( pocket_mag.empty() );
+
+            THEN( "it can contain a full clip of 9mm ammo" ) {
+                item ammo_9mm( "test_9mm_ammo", 0, full_clip_qty );
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:271]:
```suggestion
                const item ammo_9mm( "test_9mm_ammo", 0, full_clip_qty );
```

File: tests/pocket_test.cpp [hunk 19]
```diff
@@ -0,0 +1,832 @@
...
+
+            THEN( "it cannot contain items that are not ammo" ) {
+                item rag( "test_rag" );
+                expect_cannot_contain( pocket_mag, rag, "item is not an ammo",
+                                       item_pocket::contain_code::ERR_AMMO );
+            }
+        }
+
+        WHEN( "it is partly full of ammo" ) {
+            int half_clip_qty = full_clip_qty / 2;
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:292]:
```suggestion
            const int half_clip_qty = full_clip_qty / 2;
```

File: tests/pocket_test.cpp [hunk 20]
```diff
@@ -0,0 +1,832 @@
...
+            item ammo_9mm_half_clip( "test_9mm_ammo", 0, half_clip_qty );
+            pocket_mag.insert_item( ammo_9mm_half_clip );
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:294]: could be an excuse to use expect_can_insert here?

File: tests/pocket_test.cpp [hunk 21]
```diff
@@ -0,0 +1,832 @@
...
+                THEN( "item cannot be inserted into the pocket" ) {
+                    expect_cannot_insert( pocket_belt, axe, "item too big",
+                                          item_pocket::contain_code::ERR_TOO_BIG );
+                }
+            }
+        }
+
+        GIVEN( "item without BELT_CLIP flag" ) {
+            REQUIRE_FALSE( rag.has_flag( "BELT_CLIP " ) );
+            REQUIRE_FALSE( rock.has_flag( "BELT_CLIP " ) );
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:435]:
```suggestion
            REQUIRE_FALSE( rag.has_flag( "BELT_CLIP" ) );
            REQUIRE_FALSE( rock.has_flag( "BELT_CLIP" ) );
```

File: tests/pocket_test.cpp [hunk 22]
```diff
@@ -0,0 +1,832 @@
...
+// Related JSON fields:
+// "watertight"
+//
+// Functions:
+// item_pocket::watertight
+//
+TEST_CASE( "pockets containing liquids", "[pocket][watertight][liquid]" )
+{
+    // Liquids (1 charge each)
+    item ketchup( "ketchup", 0, item::default_charges_tag{} );
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:556]: the comment says 1 charge, but you are initializing it with default charges. which is it?

File: tests/pocket_test.cpp [hunk 23]
```diff
@@ -0,0 +1,832 @@
...
+{
+    item gas( "test_gas", 0, item::default_charges_tag{} );
+
+    // A potentially airtight container
+    pocket_data data_balloon( item_pocket::pocket_type::CONTAINER );
+
+    // Has capacity for several charges of gas
+    data_balloon.volume_capacity = 4 * gas.volume();
+    data_balloon.max_contains_weight = 4 * gas.weight();
+    // Avoid any length restrictions
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:663]: TODO: add exception for gases for length and max_item_volume

File: tests/pocket_test.cpp [hunk 24]
```diff
@@ -0,0 +1,832 @@
...
+            }
+
+            THEN( "item_size_modifier is the contained item volume" ) {
+                CHECK( pocket_sock.item_size_modifier() == rock.volume() );
+            }
+        }
+
+        WHEN( "pocket is full" ) {
+            // Two rocks should be enough to fill it
+            REQUIRE( pocket_sock.insert_item( item( "test_rock" ) ).success() );
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:750]: excuse to use expect_can_insert here

File: tests/pocket_test.cpp [hunk 25]
```diff
@@ -0,0 +1,832 @@
...
+            }
+
+            THEN( "item_size_modifier is zero" ) {
+                CHECK( pocket_sock.item_size_modifier() == 0_ml );
+            }
+        }
+
+        WHEN( "pocket is full" ) {
+            REQUIRE( pocket_sock.insert_item( item( "test_rock" ) ).success() );
+            REQUIRE( pocket_sock.insert_item( item( "test_rock" ) ).success() );
```
Comment (KorGgenT) [inline tests/pocket_test.cpp:787]: same as above, there's an excuse to use expect_can_insert here

**Comment by KorGgenT** [comment]
thanks for writing all these tests! i'm glad that the ret_val stuff i did for can_contain came in quite useful. i only have one more comment, that the glock test item's "magazines" do not make it so that the glock can hold any magazines after #40399 is merged, but i can see that you are not using this item as a gun in these tests at all.

File: tests/pocket_test.cpp (hunk 16 above)
Comment (wapcaplet) [reply tests/pocket_test.cpp:174]: Good call - an `is_soft` verification would help too. Will add.

File: tests/pocket_test.cpp (hunk 10 above)
Comment (wapcaplet) [reply tests/pocket_test.cpp:62]: Whoops! Nice catch. Fixed.

File: tests/pocket_test.cpp (hunk 24 above)
Comment (wapcaplet) [reply tests/pocket_test.cpp:750]: Done.

File: tests/pocket_test.cpp (hunk 20 above)
Comment (wapcaplet) [reply tests/pocket_test.cpp:294]: Done.

File: tests/pocket_test.cpp (hunk 22 above)
Comment (wapcaplet) [reply tests/pocket_test.cpp:556]: Fixed now.

File: doc/JSON_INFO.md (hunk 9 above)
Comment (wapcaplet) [reply doc/JSON_INFO.md:1891]: Good idea - I found that part in the pocket JSON loading function, but I wonder why `sqrt(2)` is used here. If I've understood right, since we're treating the pocket as a cube, the longest diagonal should be `sqrt(3)`, no?

image

Comment (wapcaplet) [reply doc/JSON_INFO.md:1891]: I've interpreted this to be a kind of "max length that fits the opening", and added tests/documentation to that effect. I can see that using `sqrt(3)` is more of an edge case for infinitely-thin items, and the need to account for an unknown (but usually significant) thickness factor makes `sqrt(2)` more sensible.

The test ensures that a 1-liter pocket with undefined max length (a 1-liter cube 10 cm on each side) allows a `10*sqrt(2) =~ 14 cm` item to fit, but an item with `longest_side: 15 cm` will not.

File: tests/pocket_test.cpp (hunk 23 above)
Comment (wapcaplet) [reply tests/pocket_test.cpp:723]: I included a related test for gases fitting through `max_item_volume` (like soft/liquid items do) but didn't get around to one for length restrictions. Maybe next PR :-)
